### PR TITLE
Issue/555 buffered action detection

### LINKED ISSUE #530: Formatting disappears when inserting space before formatted text
Originally reported in https://github.com/wordpress-mobile/AztecEditor-Android/pull/466#issuecomment-338112619

### Expected
Formatting should always be respected

### Observed
Formatting is lost for the text to the right of an inserted space

### Reproduced

1. on the editor, type "a" 
2. toggle Bold style, 
3. type "b". 
4. Put the cursor between "a" and "b" 
5. press space. Notice that "b" became unstyled.

#### Tested
[Google Pixel] on [Android 8] with [develop Aztec demo app]
tested on API 22 5.1.0, issue not observed
tested on API 17 4.2.2, issue not observed

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #530, #555

This PR implements the buffered action mechanism where events are observed and collected to detect known patterns for actions that produce more than 1 TextWatcher event per each individual user action, as otherwise implicitly expected by Aztec.

### Essentials
Instead of having all TextWatchers added directly to AztecText, we now have only one Android TextWatcher (which is implemented by AztecText itself) and we keep an internal list of other TextWatchers ourselves (`observedWatchers`). 

The original idea was to collect data from the received TextWatcher events (in their before, on, and after moments) and finally decide whether these events correspond to a known pattern such as the one described in #530 (particularly, inserting a space in a word on API26 and later is composed of 4 events: 2 deletes, and 2 inserts, instead of just 1 event that shows the space addition as can be seen on platforms other than API26+).
Then, if the pattern was detected, make up a new event that would do the same thing as in other platforms, so to standardise behaviour across Android versions, and instead send this made-up event to the regular TextWatchers that are awaiting for new events to come and be processed. All other non-corresponding events would just get processed as usual.

Unfortunately I discovered that this was impossible, as the Android’s `TextView` `TextWatcher` interface does not prevent things from being executed in any way, it just tells the listeners that something is about to happen (`beforeTextChanged`), happening (`onTextChanged`), and has just happened (`afterTextChanged`).

I decided to tweak the initial idea a little bit and, instead of trying to "_make the TextWatchers wait_",  we now only observe events happen (and dispatch them right away), and once the pattern is detected we quickly move on to operate on the content and produce the desired output.

###  Key components
- A `TextWatcherEvent` serves as a placeholder where to keep data that comes from Android’s TextWatcher interface (before, on, and after data).
- A `UserOperationEvent` is pretty much a sequence of `TextWatcherEvent`s
- A `Bucket` may contain several defined `UserOperation`s
- A specific bucket may belong to a specific platform, a range of platforms, or even a specific device model. (we now only have the implementation for API26 as described in #530 as it's the only detected case for now)
- The `ObservationQueue` is a sequence, and is filled with `TextWatcherEvents` as they come, and “piled up” if each event corresponds to an expected pattern in a known bucket, or if some divergence is found (that is, the incoming events don’t seem to match any of the expected patterns) then the queue is immediately cleared.
It is important to know that this queue is filled in with events that are very close together (under 50 milliseconds between each other), to avoid queuing up elements that are genuine user operations.
The way we do this is by comparing the time in which the events are created and check the difference between one another. We don’t use Timers nor any other elaborate structure to do that, to avoid introducing any possibility for indeterminate things to happen.
The check is also so simple, that there is no way for the queue to pile up more events than what any of the known sequences’s length in the selected buckets. For the matter of this PR which one has 1 bucket and 1 sequence implemented, this means it will only pile up up to 4 events (if they partially match the expected sequence as they come), given the pattern we observed in #530 is composed of exactly 4 events. This way, as we make sure to be looking into ALL and EVERY event, we don’t need to look into partial sequences or try to figure out or make sense out of the buffer content.

It is also important to note that the Bucket system makes the checks only available to whatever platform it was made for, so for instance running this on API17 will never enqueue anything, so no possibility to introduce new bugs on other platforms (which is something we have been running risks for before, as we would always try to find a generic way for all platforms).

### How it works

AztecText is the only TextWatcher the Android platform knows for AztecText view itself.

All of the events data is collected by it and then built into a `TextWatcherEvent` by using the builder pattern in the `afterTextChanged` method override, as we know this is the last of the 3 moments in the interface. Note that all the _before_, _on_, and _after_ events are passed to the whatever `TextWatchers` are registered to AztecText as well, as we can’t really prevent things from being executed (we can only “see” what’s going on). 

Once a `TextWatcherEvent` is built, it is immediately added to the `ObservationQueue`, which processes the event upon addition to check if a pattern corresponding to defined `UserOperation`s is building up.

The `ObservationQueue` then checks if we know this pattern or not. As soon as a queue diverges from all known patterns, it is cleared up and the queue starts to build up again with the new TextWatcher event. If it matches, it’s kept until a full pattern is found. Once found, the `UserOperationEvent` `buildReplacementEventWithSequenceData()` method is called, and with this data we end up calling the `executeEvent()` method of the registered `IEventInjector`.

The implementation of `IEventInjector` is in charge of actually operating on the content to make the desired operation happen in the way that it was expected to work.

### Other notes

- The AztecText TextWatcher implementation was caring for history keeping, we moved that into an anonymous watcher that is added in method `addHistoryLoggingWatcher`.
- Some of the code is Kotlin, some of it is still Java as I didn’t have enough expertise to cope with the complexity of this solution and deal with learning Kotlin at the same time (I tried, but was worried about time spent). I expect to move whatever is Java to Kotlin on another PR if possible. We could even wait to merge this until another Java-to-Kotlin PR gets merged against this branch as well.

### To test:
On API26+
1. on an empty AztecText instance, type “he”
2. tap on Bold
3. type “llo”
4. place the cursor right before “llo”
5. tap the space key
6. observe the format for each part of the word is kept as intended (the “he” part remains unstyled, while the “llo” part remains in Bold style).

Repeat the same on other platforms and observe the same behaviour.

cc @0nko , @daniloercoli , @maxme

### CODE REVIEW

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            if (added) {
+                processQueue()
+            }
+            return added
+        }
+    }
+
+    private fun processQueue() {
+        // here let's check whether our current queue matches / fits any of the installed buckets
+        var foundOnePartialMatch = false
```
Comment (daniloercoli) [inline aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt]: Worth to add a check on the size of the `buckets` here, and return immediately when = 0?
```
        if (buckets.size == 0)
            return
```

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+class ObservationQueue(injector: IEventInjector) : EventSequence<TextWatcherEvent>() {
+    val buckets = ArrayList<Bucket>()
+    val injector = injector
+
+    init {
+        init()
+    }
+
+    override fun add(element: TextWatcherEvent): Boolean {
+        synchronized(this@ObservationQueue) {
```
Comment (daniloercoli) [inline aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt]: Same as below, do you think  adding a check on the size of the `buckets` here, and return immediately when = 0, would improve performances on those devices that don't need additional checks?

I know, we will probably add other Buckets soon, to maybe this is not a problem.

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+import android.os.Build
+import org.wordpress.aztec.watchers.event.IEventInjector
+import org.wordpress.aztec.watchers.event.buckets.API26Bucket
+import org.wordpress.aztec.watchers.event.text.TextWatcherEvent
+import org.wordpress.aztec.watchers.event.buckets.Bucket
+
+class ObservationQueue(injector: IEventInjector) : EventSequence<TextWatcherEvent>() {
+    val buckets = ArrayList<Bucket>()
+    val injector = injector
```
Comment (daniloercoli) [inline aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt]: This can be moved to the constructors.

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+                }
+            }
+
+            eventHolder.setBeforeTextChangedEvent(observableEvent.getBeforeEventData());
+            eventHolder.setOnTextChangedEvent(observableEvent.getOnEventData());
+            eventHolder.setAfterTextChangedEvent(observableEvent.getAfterEventData());
+
+            // items in this.getSequence() should always be ITextWatcherEventComparator,
+            // but we check to avoid any potential problems
+            if (!(eventHolder instanceof ITextWatcherEventComparator
```
Comment (daniloercoli) [inline aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java]: I wonder if there is way to avoid/make this check unnecessary. Should not all of our TextWatcher events be comparable? Is an external Interface necessary?

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        // we didn't find neither a partial match nor a total match, let's just clear the queue
+        if (size > 0 && !foundOnePartialMatch) {
+            // immediately discard the queue
+            clear()
+        }
+    }
+
+    private fun init() {
```
Comment (daniloercoli) [inline aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt]: Wondering if we can remove this method and move the code  to default `init` above.

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        SpannableStringBuilder oldText = beforeData.getTextBefore();
+        oldText.insert(differenceIndex, SPACE_STRING);
+
+        replacementEvent.setAfterTextChangedEvent(new AfterTextChangedEventData(oldText));
+
+        return replacementEvent;
+    }
+
+    private void init() {
+        // here we populate our model of reference (which is the sequence of events we expect to find)
```
Comment (daniloercoli) [inline aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java]: Can't this code be moved to the constructor of the class?

**Comment by daniloercoli** [comment]
There should be some sort of recursive calls, since the demo app is not responding and crashes after while when trying to insert new text.

Steps to Repro:
1. Start the demo app
2. Point the cursor after Heading 2
3. Tap enter
4. The app is not responding and crashes after a while.

Stack Trace:
```
12-28 14:26:28.847 7537-7537/org.wordpress.aztec E/InputEventSender: Exception dispatching finished signal.
12-28 14:26:28.847 7537-7537/org.wordpress.aztec E/MessageQueue-JNI: Exception in MessageQueue callback: handleReceiveCallback
12-28 14:26:28.850 7537-7537/org.wordpress.aztec E/MessageQueue-JNI: java.lang.OutOfMemoryError: Failed to allocate a 81264648 byte allocation with 25165824 free bytes and 55MB until OOM, max allowed footprint 67968320, growth limit 100663296
                                                                         at java.util.Arrays.copyOf(Arrays.java:3260)
                                                                         at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:125)
                                                                         at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:451)
                                                                         at java.lang.StringBuilder.append(StringBuilder.java:137)
                                                                         at org.wordpress.aztec.AztecParser.withinBlock(AztecParser.kt:408)
                                                                         at org.wordpress.aztec.AztecParser.withinHtml(AztecParser.kt:387)
                                                                         at org.wordpress.aztec.AztecParser.withinHtml(AztecParser.kt:339)
                                                                         at org.wordpress.aztec.AztecParser.toHtml(AztecParser.kt:116)
                                                                         at org.wordpress.aztec.AztecText.toPlainHtml(AztecText.kt:1011)
                                                                         at org.wordpress.aztec.AztecText.toHtml(AztecText.kt:974)
                                                                         at org.wordpress.aztec.AztecText.toHtml$default(AztecText.kt:973)
                                                                         at org.wordpress.aztec.AztecText.toFormattedHtml(AztecText.kt:1015)
                                                                         at org.wordpress.aztec.History.handleHistory(History.kt:77)
                                                                         at org.wordpress.aztec.AztecText$addHistoryLoggingWatcher$historyLoggingWatcher$1.afterTextChanged(AztecText.kt:446)
                                                                         at org.wordpress.aztec.AztecText.afterTextChanged(AztecText.kt:859)
                                                                         at android.widget.TextView.sendAfterTextChanged(TextView.java:9375)
                                                                         at android.widget.TextView$ChangeWatcher.afterTextChanged(TextView.java:11917)
                                                                         at android.text.SpannableStringBuilder.sendAfterTextChanged(SpannableStringBuilder.java:1262)
                                                                         at android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:574)
                                                                         at android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:504)
                                                                         at android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:502)
                                                                         at android.text.method.QwertyKeyListener.onKeyDown(QwertyKeyListener.java:228)
                                                                         at android.text.method.TextKeyListener.onKeyDown(TextKeyListener.java:141)
                                                                         at android.widget.TextView.doKeyDown(TextView.java:7316)
                                                                         at android.widget.TextView.onKeyDown(TextView.java:7093)
                                                                         at android.view.KeyEvent.dispatch(KeyEvent.java:2688)
                                                                         at android.view.View.dispatchKeyEvent(View.java:11658)
                                                                         at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                         at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                         at android.widget.ScrollView.dispatchKeyEvent(ScrollView.java:389)
                                                                         at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                         at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                         at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                         at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                         at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                         at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                         at com.android.internal.policy.DecorView.superDispatchKeyEvent(DecorView.java:437)
                                                                         at com.android.internal.policy.PhoneWindow.superDispatchKeyEvent(PhoneWindow.java:1818)
                                                                         at android.app.Activity.dispatchKeyEvent(Activity.java:3252)
                                                                         at android.support.v7.app.AppCompatActivity.dispatchKeyEvent(AppCompatActivity.java:534)
                                                                         at android.support.v7.view.WindowCallbackWrapper.dispatchKeyEvent(WindowCallbackWrapper.java:58)
                                                                         at android.support.v7.app.AppCompatDelegateImplBase$AppCompatWindowCallbackBase.dispatchKeyEvent(AppCompatDelegateImplBase.java:316)
                                                                         at com.android.internal.policy.DecorView.dispatchKeyEvent(DecorView.java:351)
                                                                         at android.view.ViewRootImpl$ViewPostImeInputStage.processKeyEvent(ViewRootImpl.java:4714)
                                                                         at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:4586)
                                                                         at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:4128)
                                                                         at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:4181)
                                                                         at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:4147)
                                                                         at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:4274)
                                                                         at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:4155)
                                                                         at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:4331)
                                                                         at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:4128)
                                                                     	at android.view.ViewRootImpl$Inpu
12-28 14:26:28.850 7537-7537/org.wordpress.aztec D/AndroidRuntime: Shutting down VM
12-28 14:26:28.855 7537-7537/org.wordpress.aztec E/AndroidRuntime: FATAL EXCEPTION: main
                                                                   Process: org.wordpress.aztec, PID: 7537
                                                                   java.lang.OutOfMemoryError: Failed to allocate a 81264648 byte allocation with 25165824 free bytes and 55MB until OOM, max allowed footprint 67968320, growth limit 100663296
                                                                       at java.util.Arrays.copyOf(Arrays.java:3260)
                                                                       at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:125)
                                                                       at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:451)
                                                                       at java.lang.StringBuilder.append(StringBuilder.java:137)
                                                                       at org.wordpress.aztec.AztecParser.withinBlock(AztecParser.kt:408)
                                                                       at org.wordpress.aztec.AztecParser.withinHtml(AztecParser.kt:387)
                                                                       at org.wordpress.aztec.AztecParser.withinHtml(AztecParser.kt:339)
                                                                       at org.wordpress.aztec.AztecParser.toHtml(AztecParser.kt:116)
                                                                       at org.wordpress.aztec.AztecText.toPlainHtml(AztecText.kt:1011)
                                                                       at org.wordpress.aztec.AztecText.toHtml(AztecText.kt:974)
                                                                       at org.wordpress.aztec.AztecText.toHtml$default(AztecText.kt:973)
                                                                       at org.wordpress.aztec.AztecText.toFormattedHtml(AztecText.kt:1015)
                                                                       at org.wordpress.aztec.History.handleHistory(History.kt:77)
                                                                       at org.wordpress.aztec.AztecText$addHistoryLoggingWatcher$historyLoggingWatcher$1.afterTextChanged(AztecText.kt:446)
                                                                       at org.wordpress.aztec.AztecText.afterTextChanged(AztecText.kt:859)
                                                                       at android.widget.TextView.sendAfterTextChanged(TextView.java:9375)
                                                                       at android.widget.TextView$ChangeWatcher.afterTextChanged(TextView.java:11917)
                                                                       at android.text.SpannableStringBuilder.sendAfterTextChanged(SpannableStringBuilder.java:1262)
                                                                       at android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:574)
                                                                       at android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:504)
                                                                       at android.text.SpannableStringBuilder.replace(SpannableStringBuilder.java:502)
                                                                       at android.text.method.QwertyKeyListener.onKeyDown(QwertyKeyListener.java:228)
                                                                       at android.text.method.TextKeyListener.onKeyDown(TextKeyListener.java:141)
                                                                       at android.widget.TextView.doKeyDown(TextView.java:7316)
                                                                       at android.widget.TextView.onKeyDown(TextView.java:7093)
                                                                       at android.view.KeyEvent.dispatch(KeyEvent.java:2688)
                                                                       at android.view.View.dispatchKeyEvent(View.java:11658)
                                                                       at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                       at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                       at android.widget.ScrollView.dispatchKeyEvent(ScrollView.java:389)
                                                                       at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                       at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                       at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                       at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                       at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                       at android.view.ViewGroup.dispatchKeyEvent(ViewGroup.java:1827)
                                                                       at com.android.internal.policy.DecorView.superDispatchKeyEvent(DecorView.java:437)
                                                                       at com.android.internal.policy.PhoneWindow.superDispatchKeyEvent(PhoneWindow.java:1818)
                                                                       at android.app.Activity.dispatchKeyEvent(Activity.java:3252)
                                                                       at android.support.v7.app.AppCompatActivity.dispatchKeyEvent(AppCompatActivity.java:534)
                                                                       at android.support.v7.view.WindowCallbackWrapper.dispatchKeyEvent(WindowCallbackWrapper.java:58)
                                                                       at android.support.v7.app.AppCompatDelegateImplBase$AppCompatWindowCallbackBase.dispatchKeyEvent(AppCompatDelegateImplBase.java:316)
                                                                       at com.android.internal.policy.DecorView.dispatchKeyEvent(DecorView.java:351)
                                                                       at android.view.ViewRootImpl$ViewPostImeInputStage.processKeyEvent(ViewRootImpl.java:4714)
                                                                       at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:4586)
                                                                       at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:4128)
                                                                       at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:4181)
                                                                       at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:4147)
                                                                       at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:4274)
                                                                       at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:4155)
                                                                       at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:4331)
12-28 14:26:28.856 7537-7537/org.wordpress.aztec E/AndroidRuntime:     at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:4128)
                                                                       at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:4181)
                                                                       at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:4147)
                                                                       at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:4155)
                                                                       at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:4128)
                                                                       at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:4181)
                                                                       at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:4147)
                                                                       at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:4307)
                                                                       at android.view.ViewRootImpl$ImeInputStage.onFinishedInputEvent(ViewRootImpl.java:4468)
                                                                       at android.view.inputmethod.InputMethodManager$PendingEvent.run(InputMethodManager.java:2435)
                                                                       at android.view.inputmethod.InputMethodManager.invokeFinishedInputEventCallback(InputMethodManager.java:1998)
                                                                       at android.view.inputmethod.InputMethodManager.finishedInputEvent(InputMethodManager.java:1989)
                                                                       at android.view.inputmethod.InputMethodManager$ImeInputEventSender.onInputEventFinished(InputMethodManager.java:2412)
                                                                       at android.view.InputEventSender.dispatchInputEventFinished(InputEventSender.java:141)
                                                                       at android.os.MessageQueue.nativePollOnce(Native Method)
                                                                       at android.os.MessageQueue.next(MessageQueue.java:325)
                                                                       at android.os.Looper.loop(Looper.java:142)
                                                                       at android.app.ActivityThread.main(ActivityThread.java:6541)
                                                                       at java.lang.reflect.Method.invoke(Native Method)
                                                                       at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
                                                                       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
```

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt (hunk 1 above)
Comment (mzorz) [reply aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt]: I think that's what already happens in `for (bucket in buckets) { ...` when `buckets` is of size == 0; it doesn't go through the loop

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt (hunk 2 above)
Comment (mzorz) [reply aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt]: ah - I see what you mean here! it's probably best to add the check here, and this makes it not necessary to add it in the `processQueue` method, as `add()` is executed before and is the only caller of `processQueue`.
Added in 9117a31

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java (hunk 4 above)
Comment (mzorz) [reply aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java]: maybe you're right, we could make these abstract methods of the base `TextWatcherEvent` class so to make subclasses always implement that. Will check and get back here.

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt (hunk 5 above)
Comment (mzorz) [reply aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt]: Good call, I had plans for `init()` to be called in some other places but in the end it wasn't necessary. Modified this in c7f89b1

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java (hunk 6 above)
Comment (mzorz) [reply aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java]: Totally - addressed in c21ffaa.
Part of the problems I had while attempting Java/Kotlin conversions :)

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java (hunk 4 above)
Comment (mzorz) [reply aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java]: Addressed the elimination of the interface in 91f3edc - didn't make the class abstract as this is a little different situation where we're not testing `TextWatcherEvent`s against each other but rather asking the event itself to look into its inner data and determine if the data collected corresponds to its class. 

Some considerations about it: a `TextWatcherEvent` object alone doesn't necessarily know anything of this as it just holds data. But a `TextWatcherEventDeleteText` or a `TextWatcherEventInsertText` should be able to determine, by looking into the data being held within, whether this is an insertion or a deletion action happening. In this way, as we are describing an _ability_  these two classes have (but not the base class) it is that I thought an `implements interface` kind of relationship would be better suited.

I think the interface makes this a bit clearer that it is now after commit 91f3edc, but let me know your thoughts and we can either leave it as is now (with the change) or revert if you consider the prior looked better. I'd vote for the interface, but wanted to give it a try and check what you think as well 👍 - let me know your thoughts 🙇

**Comment by mzorz** [comment]
> There should be some sort of recursive calls

**Comment by mzorz** [comment]
**tl;dr**, infinite loop fixed with aa363f7.

In checking this error https://github.com/wordpress-mobile/AztecEditor-Android/pull/584#issuecomment-354296834 I realized we didn't need to keep a list of watchers ourselves anymore, but rather just keep the event list. The recursive error in history logging has been fixed by removing the `observedWatchers` array in aa363f7, now just letting the `TextView` handling the events (and attaching the `TextWatchers`) without interfering anynmore - we only now collect event data and only activate the execution of a desired event when the `ObservationQueue` is filled up with a known pattern, which is everything we need.

Ready for another pass @daniloercoli - also let me know what you think of this other comment here https://github.com/wordpress-mobile/AztecEditor-Android/pull/584/files/32fe96706287660ff03972985e5ca84dfaf903f9#r159524190

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java (hunk 4 above)
Comment (daniloercoli) [reply aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java]: Thanks Mario for taking in consideration my comments and replied with details. There is only one thing left :)
The comment in the code here https://github.com/wordpress-mobile/AztecEditor-Android/commit/91f3edc61da9d8aa6c98171ff8c6e5fa93819526#diff-faa8a86eaccb01070a180da4df06e11eR109 that still reference the Interface

**Comment by daniloercoli** [comment]
Looks Good! I'd like to get another person on the team review this PR though, since I'm not super expert of Aztec yet. Maybe @0nko can take a look?
(that's the reason why I didn't assigned it to me).

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java (hunk 4 above)
Comment (mzorz) [reply aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/known/space/API26InWordSpaceInsertionEvent.java]: 👍 addressed in 3b57fdc

File: aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt (hunk 3 above)
Comment (mzorz) [reply aztec/src/main/kotlin/org/wordpress/aztec/watchers/event/sequence/ObservationQueue.kt]: addressed the various inspector suggestions (only those that I considered applicable or made sense) in 0984aa0

**Comment by daniloercoli** [comment]
:shipit: 🚢 it
